frontend/src/metabase/query_builder/containers/QueryBuilder.jsx

```diff
@@ -147,122 +156,149 @@ const mapDispatchToProps = {
-  constructor(props, context) {
-    super(props, context);
-    this.forceUpdateDebounced = _.debounce(this.forceUpdate.bind(this), 400);
-  }
-
-  UNSAFE_componentWillMount() {
-    this.props.initializeQB(this.props.location, this.props.params);
-  }
-
-  componentDidMount() {
-    window.addEventListener("resize", this.handleResize);
-  }
+function QueryBuilder(props) {
+  const {
+    question,
+    location,
+    params,
+    fromUrl,
+    uiControls,
+    initializeQB,
+    apiCreateQuestion,
+    apiUpdateQuestion,
+    updateQuestion,
+    updateUrl,
+    locationChanged,
+    onChangeLocation,
+    setUIControls,
+    cancelQuery,
+  } = props;
+
+  // https://reactjs.org/docs/hooks-faq.html#is-there-something-like-forceupdate
+  const [, forceUpdate] = useReducer(x => x + 1, 0);
+  const forceUpdateDebounced = useMemo(() => _.debounce(forceUpdate, 400), []);
+  const timeout = useRef(null);
+
+  const previousUIControls = usePrevious(uiControls);
+  const previousLocation = usePrevious(location);
+
+  const timelineEvents = useQuestionTimelineEvents({ question });
+
+  const openModal = useCallback(modal => setUIControls({ modal }), [
+    setUIControls,
+  ]);
+
+  const closeModal = useCallback(() => setUIControls({ modal: null }), [
+    setUIControls,
+  ]);
+
+  const setRecentlySaved = useCallback(
+    recentlySaved => {
+      setUIControls({ recentlySaved });
+      clearTimeout(timeout.current);
+      timeout.current = setTimeout(() => {
+        setUIControls({ recentlySaved: null });
+      }, 5000);
+    },
+    [setUIControls],
+  );
+
+  const handleCreate = useCallback(
+    async card => {
+      const questionWithUpdatedCard = question.setCard(card);
+      await apiCreateQuestion(questionWithUpdatedCard);
+      setRecentlySaved("created");
+    },
+    [question, apiCreateQuestion, setRecentlySaved],
+  );
+
+  const handleSave = useCallback(
+    async (card, { rerunQuery = false } = {}) => {
+      const questionWithUpdatedCard = question.setCard(card);
+      await apiUpdateQuestion(questionWithUpdatedCard, { rerunQuery });
+      if (!rerunQuery) {
+        await updateUrl(questionWithUpdatedCard.card(), { dirty: false });
+      }
+      if (fromUrl) {
+        onChangeLocation(fromUrl);
+      } else {
+        setRecentlySaved("updated");
+      }
+    },
+    [
+      question,
+      fromUrl,
+      apiUpdateQuestion,
+      updateUrl,
+      onChangeLocation,
+      setRecentlySaved,
+    ],
+  );
+
+  useEffect(() => {
+    initializeQB(location, params);
+    return () => {
+      cancelQuery();
+      closeModal();
+      clearTimeout(timeout.current);
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  useEffect(() => {
+    window.addEventListener("resize", forceUpdateDebounced);
+    return () => window.removeEventListener("resize", forceUpdateDebounced);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
```

Comment:
WDYT about including `forceUpdateDebounced` here and removing the ignore comment? `forceUpdateDebounced` should never change in practice so it should be ok